src/estimagic/parameters/nonlinear_constraints.py

```diff
@@ -0,0 +1,468 @@
+
+    # ==================================================================================
+    # Process selectors for params (functional) and flat_params (indices)
+    # ==================================================================================
+
+    external_selector = _process_selector(c)  # functional selector
+
+    selection_indices, n_params = _get_selection_indices(params, external_selector)
+
+    # ==================================================================================
+    # Evaluate constraint function and selector
+    # ==================================================================================
+    selected = external_selector(params)
+
+    constraint_func = c["func"]
+
+    if constraint_eval is None:
+        constraint_eval = constraint_func(selected)
+
+    # ==================================================================================
+    # Retrieve number of constraints
+    # ==================================================================================
+    if "n_constr" in c:
+        _n_constr = c["n_constr"]
+    else:
+        constraint_value = constraint_func(external_selector(params))
+        _n_constr = len(np.atleast_1d(constraint_value))
+
+    # ==================================================================================
+    # Consolidate and transform jacobian
+    # ==================================================================================
+
+    # process numdiff_options for numerical derivative
+    options = numdiff_options.copy()
+    options.pop("lower_bounds", None)
+    options.pop("upper_bounds", None)
+
+    if "derivative" in c:
+        if not callable(c["derivative"]):
+            msg = "Jacobian of constraints needs to be callable."
+            raise ValueError(msg)
+        jacobian = c["derivative"]
+    else:
+        # use finite-differences if no closed-form jacobian is defined
+        def jacobian(p):
+            return first_derivative(
+                constraint_func,
+                p,
+                **options,
+            )["derivative"]
+
+    def _jacobian_from_internal(x):
+        """Return Jacobian of constraint at internal parameters.
+
+        The constraint function is written to be evaluated on a selection of the
+        external parameters. The optimizer, however, only works on internal parameters.
+        These can be significantly different from the external parameters, due to
+        estimagic's reparametrization features. In this function we compute the Jacobian
+        of the constraint at the internal parameters using information on the Jacobian
+        of the constraint at the selected external parameters.
+
+        """
+        params = converter.params_from_internal(x)
+        select = external_selector(params)
+        jac = jacobian(select)
+        jac_matrix = block_tree_to_matrix(jac, constraint_eval, selected)
+        jac_sparse = _extend_jacobian(jac_matrix, selection_indices, n_params)
```

Comment:
Also here, the `selection_indices` are defined quite far away. Is there away which makes it very hard to read this already complicated part of the code.